### ISSUE TITLE
Screen flicker on Android with stack navigation

### ISSUE THREAD

**Issue by jordiepasman** [react-navigation/react-navigation#10951]
### Current behavior

On Android only: every time when navigating between screens in stack navigation there will be a white flickering. 
This is much harder to see on a light background but on a black screen background it is very noticeable.

I have already found the workarounds: 
- Setting setting the dark theme (or background color) on `NavigationContainer` but we using black and white backgrounds in our application so it is still visible at one of them. 
- Also tried `animationEnabled: false` but then you won't have any animations at all
- `cardStyle` on the navigator.

There have been previous issues but I haven't really found a good solution to this yet.

This is an example on a Samsung Galaxy S20+. 
And i've added a repo to this example application.

https://user-images.githubusercontent.com/47739539/197175180-1b529a83-f895-46c7-a8da-c9ec28361d7c.mp4

### Expected behavior

No flickering and smooth navigation.

### Reproduction

https://github.com/jordiepasman/white-flickering-issue

### Platform

- [X] Android
- [ ] iOS
- [ ] Web
- [ ] Windows
- [ ] MacOS

### Packages

- [X] @react-navigation/​bottom-tabs
- [ ] @react-navigation/​drawer
- [ ] @react-navigation/​material-bottom-tabs
- [ ] @react-navigation/​material-top-tabs
- [X] @react-navigation/​stack
- [ ] @react-navigation/​native-stack

### Environment

- [] I've removed the packages that I don't use

| package                                | version |
| -------------------------------------- | ------- |
| @react-navigation/native               | ^6.0.13
| @react-navigation/bottom-tabs          |^6.4.0
| @react-navigation/material-top-tabs    | ^6.3.0
| @react-navigation/stack                | ^6.3.2
| react-native-safe-area-context         | ^4.3.1
| react-native-screens                   | ^3.18.2
| react-native-gesture-handler           |^2.5.0
| react-native-reanimated                | ^2.9.1
| react-native-tab-view                  | ^3.1.1
| react-native-pager-view                | ^5.4.25
| react-native                           | 0.70.0
| node                                   |
| yarn                            |

**Comment by raajnadar** [comment]
Set the root view colour to solve this issue, https://docs.expo.dev/versions/latest/sdk/system-ui/

**Comment by jordiepasman** [comment]
@raajnadar Thanks, but we don't use (and want) anything of Expo.

**Comment by raajnadar** [comment]
If you use a package from the expo ecosystem that doesn’t mean you need to use expo go, every package works with vanilla react native.

I tried to give you a solution, it is you who needs to find an alternative package if you don't want expo system ui

**Comment by jordiepasman** [comment]
@raajnadar Yeah I understand that and thank you for your reaction. But is the flickering not some issue of the react-navigation package instead to using workarounds? 

I commented it because the documentation says " you must have the `native expo setup`" and when I try to install it given me te same errors and I must install expo.
https://github.com/expo/expo/tree/sdk-46/packages/expo-system-ui#extra-setup 

And when I should use this package from expo can I set this on every Stack or Screen? Because we using black and white backgrounds in our application so if you not can set it dynamically it is still visible at least one of them.

**Comment by raajnadar** [comment]
> But is the flickering not some issue of the react-navigation package instead to using workarounds?

I don't think it is a workaround, we need to set the root view colour ourselves on android, on ios it is handled properly by the OS.

---

In one of my applications, the dominant colour was orange, this is what I did to solve the issue

1. In the root file inside `useEffect`, I set the root view colour as orange, (The whole application has the root view colour.

`SystemUI.setBackgroundColorAsync("orange")`

2. In our home stack (It is a tab with stack screens, they have white colour as BG)

We set the root view colour to white here so that the background of the keyboard is correct and the BG colour is white for the screens & no flicker.

---

Yes you need to do some native config so that expo packages can be used in any react native project, they also have automated installation which will make the process easier

https://docs.expo.dev/bare/installing-expo-modules/

**Comment by iway1** [comment]
> Set the root view colour to solve this issue, https://docs.expo.dev/versions/latest/sdk/system-ui/

why does this fix it just curious?

**Comment by amalmohann-dgnl** [comment]
Issue still there in "@react-navigation/native-stack": "^6.9.2",

**Comment by raajnadar** [comment]
@amalmohann-dgnl this issue is not related to the native stack

**Comment by weishengkui** [comment]
@raajnadar  actually, when i set SystemUI.setBackgroundColorAsync("white"), the issue is still exists.

**Comment by weishengkui** [comment]
@jordiepasman  @raajnadar actually, when i set SystemUI.setBackgroundColorAsync("white"), the issue is still exists.

**Comment by jordiepasman** [comment]
@zahid502 
Yes, you can do it. 
But when an app contains both dark and white elements, then there is a black flickering on the white screens.

**Comment by zahid502** [comment]
Thank you for updating @jordiepasman,

I have fixed it by using dynamic color, not static.
When there is dark mode then I have used black color matching with my dark mode screen color, 
similarly when there is light mode, then I have used white color matching with my screen color scheme.
It's worked for me.

**const themeColor = {
    ...DefaultTheme,
    colors: {
      ...DefaultTheme.colors,
      background: theme.color.background,   // matching with theme color
    },
  };**

**Note**: But it's not a proper solution of screen flicker, when there are more elements with different colors, then it will also show screen flicker on android.

**Comment by trungdvu** [comment]
Mine because of the SafeAreaProvider. So I have to change its background color to match with window color.
`
  <SafeAreaProvider style={{backgroundColor: '...'}}>
    ...
  </SafeAreaProvider>
`

**Comment by vonovak** [comment]
Another solution which will fix this and also help with theming your app is [react-native-theme-control](https://github.com/vonovak/react-native-theme-control), specifically [AppBackground component](https://github.com/vonovak/react-native-theme-control/blob/main/docs/readme-internal.md#appbackground)

**Comment by Artem-Kalugin** [comment]
In my case the problem is related to the default card interpolator, used inside react navigation for android transition animations. Same  as @jordiepasman i have a  simultaneous two theme application so i cant just set the root view color.

**I did a little tweak to default animation, to prevent flashing from being so obvious** 

Patch file located at `node_modules/@react-navigation/stack/src/TransitionConfigs/CardStyleInterpolators.tsx` with this

`+++ b/node_modules/@react-navigation/stack/src/TransitionConfigs/CardStyleInterpolators.tsx
@@ -299,7 +299,7 @@ export function forScaleFromCenterAndroid({
 
   const opacity = progress.interpolate({
     inputRange: [0, 0.75, 0.875, 1, 1.0825, 1.2075, 2],
-    outputRange: [0, 0, 1, 1, 1, 1, 0],
+    outputRange: [0, 0, 1, 1, 1, 1, 1],
   });
 
   const scale = conditional(
`

**Comment by Jayeshkarma** [comment]
Hey  just a simple solution add amination in user navigation stack 
screenOptions={{  animation:'slide_from_right' }}

**Comment by anrodrigues0** [comment]
in my case it was SafeAreaView, in which case it defines a background, it can be in the safe area provider as well
  
  <SafeAreaView style={{backgroundColor: ‘white'}}>….</SafeAreaView>

**Comment by nabilfreeman** [comment]
> Hey just a simple solution add amination in user navigation stack screenOptions={{ animation:'slide_from_right' }}

Even better, `animation: "ios"` results in a much faster slide and nice navbar transition. This removed the flash of unstyled content we were seeing, including that white flash that Expo Router introduced in version 3.

**Comment by LinhNguyen1059** [comment]
I have the same, and I fixed it by upgrading the `react-native-screens` to `3.34.0`

**Comment by erickmohor** [comment]
I'm using "@react-navigation/stack": "6.4.1" and what worked for me was set **presentation: 'transparentModal'** on the **Stack Navigator**. In my case, I'm using the Tab Navigator (3 screens) with the Stack Navigator (many screens), so I set 'presentation' to 'transparentModal' for all Stack Navigator.

Example:
```
import { createStackNavigator } from '@react-navigation/stack'

const Stack = createStackNavigator()

<Stack.Navigator screenOptions={{ presentation: 'transparentModal' }}>
    <Stack.Screen name="home" component={Home} />
</Stack.Navigator>
```

To see more details about this option: [React Navigation Doc](https://reactnavigation.org/docs/stack-navigator)

**Comment by vlummien** [comment]
For my project (Expo 52) only this helped:
Wrapping the whole project with 
`<ThemeProvider value={DarkTheme}>`

(import { DarkTheme, ThemeProvider } from '@react-navigation/native')

**Comment by moneszarrugh** [comment]
On Expo SDK 53. Tried all of the mentioned fixes, nothing worked.

**Comment by GussSoares** [comment]
> Mine because of the SafeAreaProvider. So I have to change its background color to match with window color. `<SafeAreaProvider style={{backgroundColor: '...'}}> ... </SafeAreaProvider>`

This solution work to me. Thankyou @trungdvu !

`<Stack.Navigator screenOptions={{ presentation: 'transparentModal' }}>` work too, but this feature  consume more memory because preserve the previous screen mounted.

**Comment by andveloper** [comment]
For me this only happens when I change from darkmode to light mode or vice versa. So for example starting the app with darkmode -> the navigation doesnt flicker but triggering darkmode on the device while the app is running -> flickering begins. I use a ThemeProvider but this doesnt help.

**Comment by dgreasi** [comment]
This happens to me too when using:
```
cardStyleInterpolator: CardStyleInterpolators.forHorizontalIOS,
```

I am using a NON expo react native setup

Lib versions:
```
"react-native": "0.78.3",
"@react-navigation/bottom-tabs": "6.6.1",
"@react-navigation/elements": "1.3.31",
"@react-navigation/native": "^6.1.18",
"@react-navigation/native-stack": "6.6.1",
"@react-navigation/stack": "^6.4.1",
```

For me the following worked:

- Remove all background colors from `NavigationContainer` or any wrappers in your stacks
- Set specific background colors in each specific screen in stacks, like:
```
options={{
  cardStyle: { backgroundColor: 'black' },
}}
```

You should check what background color you are using in the `NavigationContainer` and the `GestureHandlerRootView`
```
import { NavigationContainer as RNNavigationContainer } from '@react-navigation/native';
import { GestureHandlerRootView } from 'react-native-gesture-handler';
```

**Comment by MrGainn** [comment]
This solved for my, atleast make it less obvious:        <Stack>
                  <Stack.Screen
                    name="(tabs)"
                    options={{
                      headerShown: false,
                      contentStyle: { backgroundColor: "#0a0b0a" }, // ✅ background color
                    }}
                  />      </Stack>

**Comment by AryalSubin** [comment]
On Android, one workaround is to set the theme color at the top level of your router, something like this:
style={{ flex: 1, backgroundColor: "#FFF" }}

However, I think the actual issue occurs when navigating from one screen to another. If the destination screen is heavy, it takes some time to load. During this loading period, a filler screen appears (white in light mode and black in dark mode).
